doc/architecture/decisions/0002-implement-graphics-object-store.md

```diff
@@ -0,0 +1,423 @@
+# 2. implement graphics object store
+
+Date: 2024-03-28
+
+## Status
+
+Proposed
+
+## Context
+
+`gromit` currently follows a "draw-and-forget" strategy, i.e. each
+drawing operation, once completed, is drawn on a `cairo_surface_t`
+(`GromitData.backbuffer`).
+
+While this is very straightforward, it prevents the implementation of
+
+-  **dynamic content**, e.g. items that fade after a certain time and
+   disappear, or otherwise change through time
+
+-  **editing** of past content (e.g. select and move or delete an old item)
+
+Currently, a work-around exists in the form of an `aux_backbuffer`,
+which stores the state prior to the drawing operation in
+progress. This allows dynamic content during the ongoing drawing
+operation, and is used in e.g. `LINE`, `SMOOTH`, `RECT`.
+However, the drawing is split up over various callbacks (`on_press`,
+`on_motion`, and `on_release`, which complicates code and maintenance).
+
+Here, I propose to implement a **graphics object store** that contains
+the elements that have been produced so far. By re-executing these
+drawing operations, the widget is (re)painted.
+
+## Implementation
+
+A graphics object store (`objects`) is added to `GromitData`. This
+object store simply is a `GList` of `GfxObject`. Each `GfxObject` has
+the following fields:
+
+    typedef struct {
+        guint            id;                                 // unique id
+        GfxType          type;                               // type of object, e.g. Stroke, or Text
+        gboolean         dynamic;                            // content is dynamically updated
+        GromitDeviceData selected;                           // NULL, or which user selected the item
+        guint32          capabilities;                       // bit mask with capabilities (see below)
+        BoundingBox      extent;                             // the objects rectangular extent
+        void           (*exec)(GfxObject *, action, void *); // object-specific methods
+    } GfxObjectBase;
+
+This basic set of fields is extended by object-specific fields, which
+allows for an inheritance-like pattern, i.e. each specialized object
+can be downcast to the above `GfxObjectBase`.
+
+For a normal stroked path, the specific field could be:
+
+    typedef struct {
+        GfxObjectBase base;
+        GList        *coordinates;
```

Comment:
Bit of a new thought: would using https://cairographics.org/manual/cairo-Paths.html make sense or rather make this too dependent on cairo? Otherwise, I don't see us not using cairo anyway...

Reply:
I prefer to keep the GList, for two reasons:

1. the width of each path segment can vary (pressure-sensitive pen). which means the width needs to be stored with (x,y).
2. path manipulation (e.g. point reductions as I use them in SMOOTH, ORTHOGONAL etc), are very easy to apply on the GList.

Reply:
OK!